pkg/odo/cli/component/deploy.go

```diff
@@ -0,0 +1,218 @@
+		do.DockerfileBytes = dockerfileBytes
+
+		// Validate the file that was downloaded is a Dockerfile
+		err = util.ValidateDockerfile(dockerfileBytes)
+		if err != nil {
+			s.End(false)
+			return err
+		}
+
+	} else if !util.CheckPathExists(filepath.Join(do.componentContext, "Dockerfile")) {
+		s.End(false)
+		return errors.New("dockerfile required for build. No 'alpha.build-dockerfile' field found in devfile, or Dockerfile found in project directory")
+	}
+
+	s.End(true)
+
+	s = log.Spinner("Validating deployment information")
+	manifestURL := metadata.Manifest
+
+	if manifestURL == "" {
+		s.End(false)
+		return errors.New("Unable to deploy as alpha.deployment-manifest is not defined in devfile.yaml")
+	}
+
+	manifestBytes, err := util.LoadFileIntoMemory(manifestURL)
+	if err != nil {
+		s.End(false)
+		return errors.Wrap(err, "unable to download manifest from URL specified in devfile")
+	}
+	do.ManifestSource = manifestBytes
+
+	// check if manifestSource contains PORT template variable
+	// if it does, then check we have an port setup in env.yaml
+	do.DeploymentPort = 0
+	if bytes.Contains(manifestBytes, []byte("PORT")) {
+		deploymentPort, err := do.EnvSpecificInfo.GetPortByURLKind(envinfo.ROUTE)
+		if err != nil {
+			s.End(false)
+			return errors.Wrap(err, "unable to find `port` for deployment. `odo url create` must be run prior to `odo deploy`")
+		}
+		do.DeploymentPort = deploymentPort
+	}
+
+	s.End(true)
+
+	return
+}
+
+// Run has the logic to perform the required actions as part of command
+func (do *DeployOptions) Run() (err error) {
+	err = do.DevfileDeploy()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Need to use RunE on Cobra command to allow for `odo deploy` and `odo deploy delete`
+// See reconfigureCmdWithSubCmd function in cli.go
+func (do *DeployOptions) deployRunE(cmd *cobra.Command, args []string) error {
+	genericclioptions.GenericRun(do, cmd, args)
+	return nil
+}
+
+// NewCmdDeploy implements the push odo command
+func NewCmdDeploy(name, fullName string) *cobra.Command {
+	do := NewDeployOptions()
+
+	deployDeleteCmd := NewCmdDeployDelete(DeployDeleteRecommendedCommandName, odoutil.GetFullName(fullName, DeployDeleteRecommendedCommandName))
+
+	var deployCmd = &cobra.Command{
+		Use:         fmt.Sprintf("%s [command] [component name]", name),
+		Short:       "Deploy image for component",
+		Long:        `Deploy image for component`,
+		Example:     fmt.Sprintf(deployCmdExample, fullName),
+		Args:        cobra.MaximumNArgs(1),
+		Annotations: map[string]string{"command": "component"},
+		RunE:        do.deployRunE,
+	}
+	genericclioptions.AddContextFlag(deployCmd, &do.componentContext)
+
+	// enable devfile flag if experimental mode is enabled
+	deployCmd.Flags().StringVar(&do.tag, "tag", "", "Tag used to build the image")
+	deployCmd.Flags().StringSliceVar(&do.ignores, "ignore", []string{}, "Files or folders to be ignored via glob expressions.")
+
```

Comment:
```suggestion
_ = deployCmd.MarkFlagRequired("tag")
```

Reply:
We have another change ready for a default tag - will be merging soon!